crates/red_knot_python_semantic/src/types.rs

```diff
@@ -158,6 +160,35 @@ pub(crate) fn definitions_ty<'db>(
     }
 }
 
+#[derive(Debug, PartialEq, Eq, Clone, Copy, PartialOrd, Ord, Hash)]
+pub enum SupportedVersion {
```

Comment:
We need a fixed set of supported versions because otherwise the implementation becomes _much_ more complex; we can no longer just use intersections and unions, we have to implement a whole dedicated system for resolving sets of inequality constraints, like uv has, just for this one case. With a fixed set of supported versions, if we see `if sys.version_info >= (3, 11)` we can easily resolve that to `ExistsOnVersion(3.11) | ExistsOnVersion(3.12) | ExistsOnVersion(3.13)`. Without it, we have to represent `sys.version_info >= (3, 11)` as an inequality, and implement resolution for sets of inequalities.

It's _possible_, but it's a lot of extra work to take on unless we really need to. And I don't see why we'd need to; there will always be a finite and known set of Python versions supported by red-knot.

I don't think we need to support narrowing by patch version. Patch versions are not supposed to change features in ways that would require differing stub representations. Typeshed never narrows on patch versions, and neither pyright nor mypy support it. (Mypy doesn't even understand `sys.version_info` checks if they check the patch version; pyright still understands them, but treats them as if the patch version wasn't there at all.)